### DISCUSSION TITLE
dune can not work with ocaml trunk version

### DISCUSSION THREAD

**Discussion by github-neilwong** [ocaml/dune#8636] [Q&A] [answered]
Hi all
I downloaded the ocaml and it worked ok. 
ocaml
OCaml version 5.2.0+dev0-2023-04-11
Enter #help;; for help.

#exit 0;;
and i want try dune trunk version.

now this is the steps:
git clone https://github.com/ocaml/dune.git
cd dune
make release
sudo ./_boot/dune.exe install dune --prefix=/usr/local
it seems ok

but when i run make dev, it show a lot of error.

$ make dev
./_boot/dune.exe build @install
File "otherlibs/dune-rpc-lwt/examples/rpc_client/dune", line 5, characters 44-52:
5 |  (libraries dune-rpc dune-rpc-lwt csexp lwt lwt.unix))
                                                ^^^^^^^^
Error: Library "lwt.unix" not found.
-> required by
   _build/default/otherlibs/dune-rpc-lwt/examples/rpc_client/rpc_client.exe
-> required by _build/install/default/bin/dune-rpc-lwt-example-client
-> required by _build/default/dune-rpc-lwt.install
-> required by alias install
File "otherlibs/dune-rpc-lwt/src/dune", line 4, characters 38-46:
4 |  (libraries result csexp dune_rpc lwt lwt.unix unix)
                                          ^^^^^^^^
Error: Library "lwt.unix" not found.
-> required by library "dune-rpc-lwt" in
   _build/default/otherlibs/dune-rpc-lwt/src
-> required by _build/default/META.dune-rpc-lwt
-> required by _build/install/default/lib/dune-rpc-lwt/META
-> required by _build/default/dune-rpc-lwt.install
-> required by alias install
File "otherlibs/dune-build-info/src/build_info__.ml-gen", line 1:
Error: Could not find the .cmi file for interface
       otherlibs/dune-build-info/src/build_info__.ml-gen.

Can anybody help me and give some advice? thanks.

another issue:

dune --version
3.10.0-280-g34f727c

dune init project helloworld
Success: initialized project component named helloworld

cd helloworld
dune build
File "lib/helloworld.ml-gen", line 1:
Error: Could not find the .cmi file for interface lib/helloworld.ml-gen.
File "test/helloworld.ml", line 1:
Error: Could not find the .cmi file for interface test/helloworld.ml.
File "bin/main.ml", line 1:
Error: Could not find the .cmi file for interface bin/main.ml.

same problem, it means dune can not work normally.

**Comment by Alizter** [comment]
The dev target in the makefile will also build the tests which need more libraries. If you want to build those have a look at `doc/hacking.rst` for more information on development. If you just want to install dune then you won't need them.

**Reply by github-neilwong** [reply]
Thanks Alizter, i only want build dune and want helloworld run ok,

Error: Could not find the .cmi file for interface

**Reply by Alizter** [reply]
I can't help you without knowing more about the project you are trying to build. Can tell me which files you have and what is in your dune file.

**Reply by github-neilwong** [reply]
dune init project helloworld
Success: initialized project component named helloworld

cd helloworld
dune build
File "lib/helloworld.ml-gen", line 1:
Error: Could not find the .cmi file for interface lib/helloworld.ml-gen.
File "test/helloworld.ml", line 1:
Error: Could not find the .cmi file for interface test/helloworld.ml.
File "bin/main.ml", line 1:
Error: Could not find the .cmi file for interface bin/main.ml.

I just use a example to try dune, above is the steps, thanks.
My platform is ubuntu 18.04.4 LTS

cat main.ml
let () = print_endline "Hello, World!"

**Reply by Alizter** [reply]
Are you running this inside the `dune/` directory? If so, that could be causing problems. Try creating the project elsewhere.

(The explanation is roughly that dune will look for the top-most dune-project and this could be the one given by the dune repo which is surely not what you want.)

**Reply by Octachron** [reply, marked as answer]
I would also strongly advise to start with a released version of OCaml rather than with a dev version of the compiler.

In particular, there was a bug on the dev version of the compiler that might lead to "Missing cmi" errors few days ago. The bug is now fixed, but it stands to reason that if your aims is not to debug compiler bugs, you should avoid dev version of the OCaml compiler.

**Reply by Alizter** [reply]
@Octachron That's probably it, thanks for sharing!

**Comment by github-neilwong** [comment]
Thanks Octachron and Alizter, it seems have bug in ocaml trunk version, when i change to 5.1 branch, now dune work ok.

ocaml
OCaml version 5.1.1+dev0-2023-09-14
Enter #help;; for help.

#exit 0;;  
dune init project helloworld
Success: initialized project component named helloworld
dune build
_build/default/bin/main.exe
Hello, World!


when i updated to newest ocaml/dune trunk version that include bug fix, ocaml/dune also run normally.

ocaml:
commit 82d35c9e481b345127f372b38a138e6434a30524 (HEAD -> trunk, origin/trunk, origin/HEAD)
Merge: a47e554f02 53dedf18eb
Author: KC Sivaramakrishnan <kc@kcsrk.info>
Date:   Thu Sep 14 17:19:26 2023 +0530

dune:
commit 7d720b00ec7cf909a3abfb0ab010b2231e1238c9 (HEAD -> main, origin/main, origin/HEAD)
Author: Ali Caglayan <alizter@gmail.com>
Date:   Thu Sep 14 16:28:30 2023 +0100

ocaml --version
The OCaml toplevel, version 5.2.0+dev0-2023-04-11
dune --version
3.10.0-295-g7d720b0
